sdk/nodejs/automation/cmd.ts

```diff
@@ -37,26 +44,122 @@ export class CommandResult {
     }
 }
 
-const unknownErrCode = -2;
+export interface PulumiOptions {
+    version?: semver.SemVer;
+    root?: string;
+}
 
-/** @internal */
-export async function runPulumiCmd(
+export class Pulumi {
+    private constructor(readonly command: string, readonly version: semver.SemVer) {}
+
+    static async get(opts?: PulumiOptions): Promise<Pulumi> {
+        const command = opts?.root ? path.resolve(path.join(opts.root, "bin/pulumi")) : "pulumi";
+
+        const { stdout } = await exec(command, ["version"]);
+
+        const version = semver.parse(stdout) || semver.parse("3.0.0")!;
+        if (opts?.version && version.compare(opts.version.toString()) < 0) {
+            throw Error(`${command} version ${version} does not satisfy version ${opts.version}`);
+        }
+        return new Pulumi(command, version);
+    }
+
+    static async install(opts?: PulumiOptions): Promise<Pulumi> {
+        const optsWithDefaults = withDefaults(opts);
+        try {
+            return await Pulumi.get(optsWithDefaults);
+        } catch (err) {
+            // ignore
+        }
+
+        if (process.platform === "win32") {
+            await Pulumi.installWindows(optsWithDefaults);
+        } else {
+            await Pulumi.installPosix(optsWithDefaults);
```

Comment:
Not sure where the right place is, but I think we might want some error handling if either the script download, or the script itself fails, to at least provide a nice error message to the user?  Or is that somehow included with the nodejs runtime, and we have to do nothing?  (I'm not super familiar with typescript, so happy to be told otherwise)

Reply:
If the download or install fail, it will throw an exception that the user can catch.

For failed downloads we get something like `HTTPError: Response code 404 (Not Found)`

When the script fails, we get for example:

```
CommandError: code: -2
   stdout:
   stderr: Command failed with exit code 1: /bin/sh /var/folders/4q/3skx3qb5191bjhkchcbbrmmm0000gn/T/tmp-25358-s8q9bWmq4iDU --no-edit-path --install-root /tmp/plop --version 3.9.1

  We're sorry, but it looks like something might have gone wrong during installation.
  If you need help, please join us on https://slack.pulumi.com/
  error: /tmp/plop/bin/pulumi already exists and is a directory, refusing to proceed.
```